### ISSUE TITLE
Device (node, mqtt) disconnects during direct method execution

### ISSUE THREAD

**Issue by tameraw** [Azure/azure-iot-sdk-node#11]
_From @bmrpatel on November 22, 2016 12:55_

**Platform: Linux, Nodejs 7.1, SDK v 1.1 MQTT**

_**Scenario:**_ Device is connected and method gets invoked. Direct Method is executed. I am using the sample code with the following changes to test async execution. Using DeviceExplorer to invoke method. As soon as `onWriteLine` is called, it gets disconnected and get the error on the DeviceExplorer side. 

```
client.onDeviceMethod('writeLine', (req, res) => {
            console.log('writeLine method req received!');
            setTimeout(() => {
                onWriteLine(req, res)
            }, 1000)
        });
``` 
![image](https://cloud.githubusercontent.com/assets/5115789/20524367/418c6542-b088-11e6-9e2c-29ddab206462.png)

may be related to #980

_Copied from original issue: Azure/azure-iot-sdks#981_

**Comment by tameraw** [comment]
_From @bmrpatel on November 22, 2016 15:42_

**DEBUG Output**
![image](https://cloud.githubusercontent.com/assets/5115789/20525227/7a89142c-b08c-11e6-9905-40ea49eaa0fd.png)

**Comment by tameraw** [comment]
_From @bmrpatel on November 22, 2016 22:16_

I noticed the pattern and I see that it disconnects when it happens. I am making a web request call to get some data which is completely unrelated to the iot client. When this request is taking long time to receive a response, it disconnects and following debug output is produced. I am guessing mqtt is doing some health check in background. This background check may be timing out and hence disconnecting since other task is taking long time. Please let me know if you need help reproducing this issue.

` mqtt-common Device connection to the server has been closed`

**Comment by tameraw** [comment]
_From @bmrpatel on November 24, 2016 18:17_

I get the following error:
![image](https://cloud.githubusercontent.com/assets/5115789/20607654/2ba8ab34-b248-11e6-9889-2022fab57e85.png)

**Comment by tameraw** [comment]
Investigating.

**Comment by tameraw** [comment]
_From @bmrpatel on December 5, 2016 14:15_

**UPDATE:** @pierreca I get the same error when I set a break point to debug after it connects to the hub. I try to evaluate for few seconds to minute and see the same error in console. 

It is weird that it is not calling/emitting disconnect event on the client. If I get that then at least I can handle that and try to connect again. What is the policy to handle this kind of close events?

I am wondering if this is deployed to production by anyone, are they not seeing this issue? Or it is being handled and I am unaware of the way to do that! 

Thanks a lot..

**Comment by tameraw** [comment]
_From @bmrpatel on December 8, 2016 14:2_

@pierreca, @tameraw  I ran another test. I modified one of the device to use AMQP (2 line change). Otherwise both devices are running the same program. MQTT device disconnected the same day after couple of hours and AMQP device is still connected after 3 days and responding from cloud.

**Comment by tameraw** [comment]
_From @sergaz-msft on January 12, 2017 12:8_

Hi @bmrpatel thanks for the details. I will try to reproduce the same on my Environment. I am also working on #980 so we will check if both issues are related or not.

**Comment by sergaz-msft** [comment]
Hi @bmrpatel, like we agreed on #12 please change Line 57 on mqtt.js and let me know if it fixes this issue as well.

**Comment by pierreca** [comment]
link to fix: https://github.com/Azure/azure-iot-sdk-node/commit/a46c4a1de0008e9c61f0b86bce8d54965b438914

will be closed when we release.

**Comment by bmrpatel** [comment]
Hi @pierreca, I tried this change. Now it reconnects fine in case of close. However, it does fail to respond to the method. Is there a way to respond to the same method request after reconnecting, in case of disconnect? Or can you handle this and make sure it does not disconnect when busy?

Thanks

**Comment by pierreca** [comment]
@bmrpatel  here's my understanding of the problem please let me know if you agree or disagree:

Method calls do not persist over disconnections, it's a current limitation of the feature at the service level (ie there's nothing the SDK can do to circumvent that: we are looking at ways to improve the service and connectivity to remove this limitation, but for now it's there).

We could try and figure out a logic around how to "wait" before we disconnect to renew the token if a method is expecting a response, but we have no control over how much time crafting this response takes (since it's user code) so ultimately there would be cases where this logic would fail (token would expire before the method response is sent). 

Since disconnection can happen at any time and are not necessarily related to renewing the token (could be network failure, etc), redundancy/safety should be built on the service calling the device method in the first place to make sure the method was correctly executed or not on devices.

Does it make sense?

**Comment by bmrpatel** [comment]
@tameraw I agree, thanks for the clarifications.

**Comment by pierreca** [comment]
The fix has been released to npm with packages version 1.1.4:
- https://www.npmjs.com/package/azure-iot-device-mqtt
- https://www.npmjs.com/package/azure-iot-device
